### PR TITLE
Duplicator

### PR DESCRIPTION
**Pull request by Shadowmech88**
duplicator_complete
http://puu.sh/tq5VT/cf64e838ae.webm

Adds the duplicator, a machine which will duplicate anything placed into one pod, with the duplicate in the other pod possessing all the same contents and var values as the original.

Adds a `duplicate()` proc, which creates a new instance of the datum it's called on and then copies its vars over to the duplicate. Dumb copying of references tends to cause really bad unforeseen stuff, so the proc can be overridden for individual types in order to specifically copy the needed references in a more controlled manner, such as I've done for humans in needing to copy their organs and organ damages.

Console sprite taken from Europa Station.

### CODE REVIEW

**Comment by jakmak5** [comment]
So xenoarch artifact?

**Comment by TheFriggenSawbones** [comment]
this is super cool

**Comment by ScrambledSteak** [comment]
If you duplicate a person, ghosts should be pinged to ask if anybody wants to occupy the new body. Have wacky clone antics straight out of a saturday morning cartoon.

**Comment by Shadowmech88** [comment]
It's not obtainable outside of adminbus *yet*. I'll do that in a separate PR.

**Comment by kol1th** [comment]
Would a duplicated mob retain it's allegiances?

**Comment by Shadowmech88** [comment]
It depends on how allegiances are stored. Without any modification, the duplication process directly copies all non-reference vars and all lists that don't contain references, as well as the contents list.

**Comment by kol1th** [comment]
Hm, alright. 
 
Also, +1 to the idea of players being polled into duplicated humans and whatnot, so they're just not catatonic messes. 

I would say to either have it in a new vault with some LORE, or a xenoarch find. I like the vault idea more, since it would probably be used/found more. 

Also, maybe request some sprites for the left and right sides so they match up with the middle piece?

**Comment by Shadowmech88** [comment]
I think the pods look fine as-is, but if anyone wants to volunteer sprites, feel free.

**Comment by jknpj** [comment]
@kol1th l think it is too strong to be anything but an artifact/bus. I mean come on you can duplicate almost ANYTHING with it FOR FREE

**Comment by Zth--** [comment]
I'd like to see it as a machine that requires phazon to build. I don't think it's overpowered to be fair.

**Comment by ultraplen** [comment]
Does it duplicate an items contents as well?

**Comment by Shadowmech88** [comment]
Yes.

**Comment by Aurx** [comment]
A: While fun, this is going to be OP as balls.  Imagine telescibombing with infinite TTVs, or printing hundreds of caplasers, or a dozen xeno queens.  
B: This needs failsafes so that it doesn't even attempt to duplicate something it'd risk breaking.  Even if the player who printed the thing understands it might break, any schmuck who picks it up could easily assume it SHOULD work fine and create a bug report that'd be a nightmare to reproduce.  Or a runtime that'd be utterly inexplicable.

File: code/modules/mob/living/carbon/human/human.dm [hunk 1]
```diff
@@ -1669,3 +1669,54 @@
 	if(pain_numb)
 		return FALSE
 	return TRUE
+
+/mob/living/carbon/human/duplicate(atom/destination)
```
Comment (ComicIronic) [inline code/modules/mob/living/carbon/human/human.dm:1673]: This version does not return dupe.

File: code/modules/organs/organ.dm [hunk 2]
```diff
@@ -238,3 +238,39 @@
...
+				continue
+			if(istype(vars[x], /list))
+				var/list/OL = vars[x]
+				var/list/DL = dupe.vars[x]
+				if(!DL || !OL)
+					if(!DL)
+						DL = list()
+					if(!OL)
+						OL = list()
+				if(OL && DL)
```
Comment (ComicIronic) [inline code/modules/organs/organ.dm:262]: This line is unnecessary - you've ensured the condition is met above, and Byond is single threaded, so no other code can execute between the other code and this.

File: code/modules/organs/organ.dm [hunk 3]
```diff
@@ -238,3 +238,39 @@
...
+						if(istext(y) || isnum(y))
+							if(!isnum(y) && OL[y])
+								DL[y] = OL[y]
+							else
+								for(var/i = 1, i <= OL.len, i++)
+									if(OL[i] == y)
+										DL.Add(y)
+										break
+						else
+							break
```
Comment (ComicIronic) [inline code/modules/organs/organ.dm:273]: Why do you break on the first object in the list?

File: code/modules/organs/organ.dm [hunk 4]
```diff
@@ -238,3 +238,39 @@
...
+	var/datum/organ/dupe = new type()
+
+	for(var/x in vars)
+		if(!exclude.Find(x)) // Important!
+			if(x == "owner")
+				dupe.owner = destination
+				continue
+			if(istype(vars[x], /datum))
+				continue
+			if(istype(vars[x], /list))
```
Comment (ComicIronic) [inline code/modules/organs/organ.dm:254]: See if you can make this behaviour part of /list/duplicate().

File: code/modules/research/xenoarchaeology/artifact/artifact_duplicator.dm [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+	update_icon()
+	if(anchored && !deployed)
+		deploy_pods()
+	else if(!anchored && deployed)
+		retract_pods()
+
+/obj/machinery/duplicator/proc/deploy_pods()
+	desc = "It's some kind of large computer."
+	visible_message("The disks detach from the sides of \the [src] and unfold into large pods.")
+	leftpod.forceMove(locate(x-1,y,z))
```
Comment (ComicIronic) [inline code/modules/research/xenoarchaeology/artifact/artifact_duplicator.dm:44]: Use `get_step()`, and have it be based on direction variables on the object - it would be useful to be able to map these for vaults where they aren't left-to-right.

File: code/modules/research/xenoarchaeology/artifact/artifact_duplicator.dm [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+	if(..() || !anchored)
+		return 1
+	interact(user)
+
+/obj/machinery/duplicator/interact(mob/user)
+	if(!leftpod || !rightpod)
+		visible_message("\The [src] buzzes \"<span class='warning'>ERROR: Both pods not present.</span>\"")
+		flick("duplicator_denied", src)
+		return
+	if(!leftpod.get_object() || !rightpod.get_object())
```
Comment (ComicIronic) [inline code/modules/research/xenoarchaeology/artifact/artifact_duplicator.dm:67]: You call `leftpod.get_object()` about 4 times in this code block, and similarly for the right pod - just store the result of the first call in a var and check that.

File: code/modules/research/xenoarchaeology/artifact/artifact_duplicator.dm [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+		return
+	duplicating = TRUE
+	update_icon()
+	A.forceMove(sourcepod)
+	sourcepod.icon_state = "borgcharger1(old)"
+	sourcepod.density = 1
+	destpod.icon_state = "borgcharger1(old)"
+	destpod.density = 1
+	A.duplicate(destpod)
+	spawn(dupe_time)
```
Comment (ComicIronic) [inline code/modules/research/xenoarchaeology/artifact/artifact_duplicator.dm:108]: Because there's no code after this spawn, it's better to set `waitfor = 0` at the start of the proc and use a sleep instead - it helps with horizontal readability.

**Review by ComicIronic — CHANGES_REQUESTED** [review]

File: code/game/objects/items.dm [hunk 8]
```diff
@@ -1092,4 +1092,69 @@ var/global/list/image/blood_overlays = list()
...
 		return new shrapnel_type(src)
 	else
-		return 0
\ No newline at end of file
+		return 0
+
+/obj/item/duplicate(atom/destination)
+	var/obj/item/dupe = ..()
+
+	if(iscarbon(loc) && iscarbon(destination))
```
Comment (ComicIronic) [inline code/game/objects/items.dm:1100]: Why is this required? Shouldn't human and monkey code just be duplicating clothes items itself? This is a serious break of encapsulation.

File: code/modules/organs/organ.dm (hunk 3 above)
Comment (Shadowmech88) [reply code/modules/organs/organ.dm:273]: The first object that isn't text or num, to prevent it trying to copy lists of references.
Though I suppose there's no reason to break the loop, just in case it happens to somehow be a list of both text/nums and references.

File: code/modules/research/xenoarchaeology/artifact/artifact_duplicator.dm (hunk 5 above)
Comment (Shadowmech88) [reply code/modules/research/xenoarchaeology/artifact/artifact_duplicator.dm:44]: >and have it be based on direction variables on the object

That would look dumb because the console doesn't have any other directional sprites.

Comment (ComicIronic) [reply code/modules/research/xenoarchaeology/artifact/artifact_duplicator.dm:44]: It doesn't matter - the cloning console doesn't have directional sprites, but you can still map it how you like.

File: code/game/objects/items.dm (hunk 8 above)
Comment (Shadowmech88) [reply code/game/objects/items.dm:1100]: The basic parent `duplicate()` proc duplicates contents, since that's relatively safe.
Since clothing worn by humans and monkeys is placed in the mob's contents, duplicating a human will duplicate all the human's clothes and place them in the duplicate's contents.

For the purpose of assigning each piece of clothing to the correct inventory slots, putting this code in `human/duplicate()` wouldn't work because the code would have no way to guarantee which original item each duplicate item came from, and so no way to confirm which inventory slot the duplicate should be placed in.

File: code/modules/organs/organ.dm (hunk 2 above)
Comment (Shadowmech88) [reply code/modules/organs/organ.dm:262]: I added this because, in spite of the above code, I was getting bad index/cannot read from list errors. But I'll remove it and see what happens.

File: code/game/objects/items.dm (hunk 8 above)
Comment (ComicIronic) [reply code/game/objects/items.dm:1100]: Aren't the equip vars covered by a var loop anyways?

Comment (Shadowmech88) [reply code/game/objects/items.dm:1100]: No, because the equip vars are references, and no references are copied by the default `duplicate()` proc.

Comment (ComicIronic) [reply code/game/objects/items.dm:1100]: Then I would recommend having each atom be able to provide a list of objects using a proc that should not be duplicated from its contents, and checking that in the parent proc. Then human and monkey code can handle inventory itself.

File: code/modules/organs/organ.dm (hunk 4 above)
Comment (PJB3005) [reply code/modules/organs/organ.dm:254]: Can you ever  define procs on lists?

**Comment by Durgen** [comment]
checkem

File: code/modules/organs/organ.dm (hunk 4 above)
Comment (RemieRichards) [reply code/modules/organs/organ.dm:254]: you can't.

**Comment by Blithering** [comment]
1% chance of the duplicated object being evil and having a tiny goatee?

File: code/modules/research/xenoarchaeology/artifact/artifact_duplicator.dm [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+		sourcepod.density = 0
+		destpod.icon_state = initial(destpod.icon_state)
+		destpod.density = 0
+		A.forceMove(get_turf(sourcepod))
+		for(var/atom/movable/I in destpod.contents)
+			I.forceMove(get_turf(destpod))
+		duplicating = FALSE
+		update_icon()
+
+/datum/proc/duplicate(atom/destination)
```
Comment (clusterfack) [inline code/modules/research/xenoarchaeology/artifact/artifact_duplicator.dm]: This is literally the worst idea anyone has ever had

Comment (Exxion) [reply code/modules/research/xenoarchaeology/artifact/artifact_duplicator.dm:122]: Not as bad as 25 force toolboxes

Comment (clusterfack) [inline code/modules/research/xenoarchaeology/artifact/artifact_duplicator.dm]: No, this is infinitely worse

**Comment by gbasood** [comment]
This is way, way too strong.

**Comment by clusterfack** [comment]
You cannot, and should not ever programmatically copy over variables from one object to another.

**Comment by Shadowmech88** [comment]
Why not?

**Comment by clusterfack** [comment]
Because it can and will break in unexpected ways that you will never be able to foresee and prevent.

**Comment by gbasood** [comment]
One example would be garbage collection checks that happen where two objects reference one another would no longer work.
If  thing A derefered thing B and had thing B dereference thing A when thing A goes away, a duplicate of thing B wouldn't dereference thing A if thing A went away because thing A wouldn't have a reference to tell it to dereference it

**Comment by Exxion** [comment]
Also, list aliasing is the devil

**Comment by Shadowmech88** [comment]
But the default duplicate proc doesn't copy references.

**Comment by clusterfack** [comment]
So it'll runtime when for example it thinks it is running a process and holding a variable, but the variable it should be referencing isn't there.

You've made a runtime machine.

**Comment by clusterfack** [comment]
To make a machine with this kind of behavior you have to do it by hand for each object you want to come out 'identically', and for any other object just create a new one instead.

**Comment by gbasood** [comment]
Issues that break the code aside, a machine that dupes things is kind of dumb
Comsumable dupe of one item? *Maybe* better

**Comment by Shadowmech88** [comment]
I plan on finishing this, but not right now and when I do it'll require a major rework as to the method, so no reason to leave it open for now.
